kernel/src/platform/scheduler_timer.rs

```diff
@@ -0,0 +1,210 @@
+    ///
+    /// This must set a timer for an interval as close as possible to the given
+    /// interval in microseconds. Interrupts do need to be enabled.
+    ///
+    /// Callers can assume at least a 24-bit wide clock. Specific timing is
+    /// dependent on the driving clock. For ARM boards with a dedicated SysTick
+    /// peripheral, increments of 10ms are most accurate thanks to additional
+    /// hardware support for this value. ARM SysTick supports intervals up to
+    /// 400ms.
+    fn start(&self, us: u32);
+
+    /// Reset the scheduler timer.
+    ///
+    /// This must reset the timer, and can safely disable it and put it in a low
+    /// power state. Calling any function other than `start()` immediately after
+    /// `reset()` is invalid.
+    ///
+    /// Implementations should disable the timer and put it in a lower power
+    /// state, but this will depend on the hardware and whether virtualization
+    /// layers are used.
+    fn reset(&self);
+
+    /// Arm the SchedulerTimer timer and ensure an interrupt will be generated.
+    ///
+    /// The timer must already be started by calling `start()`. This function
+    /// guarantees that an interrupt will be generated when the already started
+    /// timer expires. This interrupt will preempt the running userspace
+    /// process.
+    fn arm(&self);
+
+    /// Disarm the SchedulerTimer timer indicating an interrupt is no longer
+    /// required.
+    ///
+    /// This does not stop the timer, but indicates to the SchedulerTimer that
+    /// an interrupt is no longer required (i.e. the process is no longer
+    /// executing). By not requiring an interrupt this may allow certain
+    /// implementations to be more efficient by removing the overhead of
+    /// handling the interrupt. The implementation may disable the underlying
+    /// interrupt if one has been set, depending on the requirements of the
+    /// implementation.
+    fn disarm(&self);
+
+    /// Return the number of microseconds remaining in the process's timeslice.
+    fn get_remaining_us(&self) -> u32;
+
+    /// Check if the process timeslice has expired.
+    ///
+    /// Returns `true` if the timer has expired since the last time this
+    /// function or `set_timer()` has been called. This function may not be
+    /// called after it has returned `true` for a given timeslice until
+    /// `set_timer()` is called again (to start a new timeslice). If `expired()`
+    /// is called again after returning `true` without an intervening call to
+    /// `set_timer()`, the return the return value is unspecified and
+    /// implementations may return whatever they like.
+    ///
+    /// The requirement that this may not be called again after it returns
+    /// `true` simplifies implementation on hardware platforms where the
+    /// hardware automatically clears the expired flag on a read, as with the
+    /// ARM SysTick peripheral.
+    fn expired(&self) -> bool;
```

Comment:
I might bikeshed this to `is_expired`; the return type is clear, but the name sounds closed to the use of a function that would be called on the implementing object when the timer expires

Reply:
I could see `has_expired()`, since it can only be called once. `expired()` is good too though.